HPCC-28544 DFUServer crashes when source file is not distributed file.

AttilaVamos opened the pull request:
Check distributedSource/distributedTarget value before use.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith commented:
Not sure this is the right fix, but could be a short term workaround.

If the source or the target (set via setSource/setTarget) was a IFileDescriptor not a IDistributedFile, it implies it was not a published logical file, e.g. because it's a file on a LZ. ( it looks like it's also true via other routes, e.g. IDistributeFileSystem::replicate ).

I think the copy API can still be used in some of these cases, it would mean getAPICopyClient would have to determine the plane(s) involved via other means (e.g. extracting and setting as members during setSource/setTarget)

I'll defer to @shamser though.

ghalliday commented on dali/ft/filecopy.cpp:
```diff
@@ -1522,6 +1522,9 @@ IAPICopyClient * FileSprayer::getAPICopyClient()
 
+    if (!distributedSource)
+        return nullptr;
+
```
I think this change is good.  There should also be checks that sourcePlane and targetPlane are not null before they are derefenced.

AttilaVamos replied:
Two more checks added

ghalliday commented:
I also agree with Jake, that it should be possible to use the file copy for spray/despray, but a quick fix is needed.  I think this should probably go into an rc2 since it is a significant regression.

AttilaVamos commented:
I agree with both of you, this is a quick fix only and we need a better approach.

jakesmith reviewed (approved):
Could mege as is, but it is probably worth adding the sourcePlane and targetPlane checks Gavin pointed out in this PR too.

ghalliday commented:
@AttilaVamos I think you need both checks.  Please can you add them back and repush.

AttilaVamos committed:
HPCC-28544 DFUServer crashes when source file is not distributed file.

Check distributedSource/distributedTarget value before use.

Check sourcePlane/targetPlane value before use.

Tested manually.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@ghalliday squashed.
